### PR TITLE
azuread_group - support for the `owners` property

### PR DESCRIPTION
**Pull request by tiwood**
(1/2 of #36)

This adds a new resource to manage Azure AD Group Owners with Terraform. 
Additionally this adds support for `owners` in the resource azuread_group.

- [x] New resource 'azuread_group_owner'
- [x] New attribute 'owners' for resource 'azuread_group'
- [x] Add tests
- [x] Add documentation
- [ ] azuread_group: If `owners` is not set, it should leave existing owners as is.

## Example Usage

```hcl
resource "azuread_group" "my_group" {
  name = "MyGroup"
}

data "azuread_user" "my_user" {
  user_principal_name = "johndoe@hashicorp.com"
}

resource "azuread_group_owner" "default_owner" {
  group_object_id = "${azuread_group.my_group.id}"
  owner_object_id = "${data.azuread_user.my_user.id}"
}
```

## Example for azuread_group

```hcl
resource "azuread_group" "my_group" {
  name     = "my group"
  owners  = ["9c9b3992-78ac-11e9-a3ef-a31d2f3e9d2c"] 
}
```

## Argument Reference

The following arguments are supported:

* `group_object_id` - (Required) The object id of the Azure AD Group where the Owner should be added.
* `owner_object_id` - (Required) The object id of the Azure AD User you want to add as Owner.

## Caveats

Azure requires at least one owner per Azure AD group. That means the destruction of 'azuread_group_owner' resources will fail if no additional owner is present on the group.

### CODE REVIEW

**Comment by tiwood** [comment]
Related: #36

**Comment by tiwood** [comment]
@katbyte, @tombuildsstuff, do you guys have any ideas how we could tackle the issue of the 'not deletable' (see known issues above) group owners?

BTW you have the same behaviour you can see in the Azure Portal.
1. Create a group
2. Add a new owner
3 Try to remove the owner (this fails if its the last owner of the group)

Should we leave this up to the User?

**Comment by twendt** [comment]
Would it be an option to not implement it as a separate resource but instead integrate it into the group resource? There you could always pass the full list of owners in and not allow the list to be empty.

**Comment by tiwood** [comment]
@twendt, I think this is the right way to do it. I'm going to look into it.
Thanks.

**Comment by katbyte** [comment]
@tiwood & @twendt,

There are reasons to have it as a separate resource (you can separate out creating & managing the owners) & within the resource (central location, ability to explicitly define who the owners should be)

And as such we should support both.

as to the issue of not being able to remove the last user, I say we fail destruction/update and leave that up to the user as theres not much we can do about it.

**Comment by tiwood** [comment]
I will look into this in the next days.

File: azuread/resource_group_owner.go [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+	tenantID := client.TenantID
+
+	ownerGraphURL := fmt.Sprintf("https://graph.windows.net/%s/directoryObjects/%s", tenantID, ownerID)
+
+	properties := graphrbac.AddOwnerParameters{
+		URL: &ownerGraphURL,
+	}
+
+	_, err := client.AddOwner(ctx, groupID, properties)
+	if err != nil {
```
Comment (katbyte) [inline azuread/resource_group_owner.go]: _minor_ these two lines could be merged:
```
	if _, err := client.AddOwner(ctx, groupID, properties); err != nil {
```

File: azuread/resource_group_owner.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+		user, _ := owners.Value().AsUser()
+		if user != nil {
+			if *user.ObjectID == ownerID {
+				ownerObjectID = *user.ObjectID
+				break
+			}
+		}
+
+		err = owners.NextWithContext(ctx)
+		if err != nil {
```
Comment (katbyte) [inline azuread/resource_group_owner.go]: These two lines could be merged

File: website/docs/r/group_owner.markdown [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  group_object_id = "${azuread_group.my_group.id}"
+  owner_object_id = "${data.azuread_user.my_user.id}"
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `group_object_id` - (Required) The object id of the Azure AD Group where the Owner should be added.
```
Comment (katbyte) [inline website/docs/r/group_owner.markdown]: Could we mark this as force new
```suggestion
* `group_object_id` - (Required) The object id of the Azure AD Group where the Owner should be added. Changing this forces a new resource to be created.
```

File: website/docs/r/group_owner.markdown [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+* `owner_object_id` - (Required) The object id of the Azure AD User you want to add as Owner.
```
Comment (katbyte) [inline website/docs/r/group_owner.markdown]: and here
```suggestion
* `owner_object_id` - (Required) The object id of the Azure AD User you want to add as Owner. Changing this forces a new resource to be created.
```

**Comment by tiwood** [comment]
I'm almost done with the changes, but I'm currently trying to figure out something.

Is it possible to suppress a diff if the ordering of a list has changed?

```
testing.go:538: Step 0 error: After applying this step, the plan was not empty:

        DIFF:

        UPDATE: azuread_group.test
          owners.0: "0911ca82-5526-4d57-bfd6-88cb71057df4" => "dd447a48-663e-4b2a-9d0a-db9db3287d72"
          owners.1: "dd447a48-663e-4b2a-9d0a-db9db3287d72" => "0911ca82-5526-4d57-bfd6-88cb71057df4"
```

The schema property is defined as follows:
```
Schema: map[string]*schema.Schema{
			"name": {
				Type:         schema.TypeString,
				Required:     true,
				ForceNew:     true,
				ValidateFunc: validation.NoZeroValues,
			},
			"owners": {
				Type:     schema.TypeList,
				Optional: true,
				MinItems: 1,
				MaxItems: 100, //Group owners are maxed out at 100
				Elem: &schema.Schema{
					Type:         schema.TypeString,
					ValidateFunc: validate.UUID,
				},
			},
		},
```

File: website/docs/r/group_owner.markdown (hunk 4 above)
Comment (tiwood) [reply website/docs/r/group_owner.markdown]: Changed as requested.

File: website/docs/r/group_owner.markdown (hunk 3 above)
Comment (tiwood) [reply website/docs/r/group_owner.markdown]: Changed as requested.

File: azuread/resource_group_owner.go (hunk 2 above)
Comment (tiwood) [reply azuread/resource_group_owner.go]: Changed as requested.

File: azuread/resource_group_owner.go (hunk 1 above)
Comment (tiwood) [reply azuread/resource_group_owner.go]: Changed as requested.

File: azuread/resource_group.go [hunk 5]
```diff
@@ -83,11 +93,22 @@ func resourceGroupCreate(d *schema.ResourceData, meta interface{}) error {
 	if v, ok := d.GetOk("members"); ok {
 		members := tf.ExpandStringSlicePtr(v.(*schema.Set).List())
 
+		// we could lock here against the group ember resource, but the should not be used together (todo conflicts with at a resource level?)
```
Comment (mbfrahry) [inline azuread/resource_group.go]:
```suggestion
		// we could lock here against the group member resource, but they should not be used together (todo conflicts with at a resource level?)
```

File: azuread/resource_group.go [hunk 6]
```diff
@@ -83,11 +93,22 @@ func resourceGroupCreate(d *schema.ResourceData, meta interface{}) error {
...
 		if err := graph.GroupAddMembers(client, ctx, *group.ObjectID, *members); err != nil {
 			return err
 		}
 	}
 
+	// Add owners if specified
+	if v, ok := d.GetOk("owners"); ok {
+		members := tf.ExpandStringSlicePtr(v.(*schema.Set).List())
+
+		// we could lock here against the group owner resource, but the should not be used together (todo conflicts with at a resource level?)
```
Comment (mbfrahry) [inline azuread/resource_group.go]:
```suggestion
		// we could lock here against the group owner resource, but they should not be used together (todo conflicts with at a resource level?)
```

**Comment by katbyte** [comment]
Closes #36

**Pull request merged**